### PR TITLE
Cdn

### PR DESCRIPTION
**Pull request by coreyreichle**
Initial start for CDN documentation, re: issue #467

### CODE REVIEW

**Comment by nightpool** [comment]
what value does this documentation add? Is this a WIP or the finished state?

(From the description I assume it's a work in progress, but on #467 you mentioned it was ready for review)

**Comment by coreyreichle** [comment]
The Cloudflare section is complete.  I'm sure there are other CDNs that are used, however, I have no experience with them fronting Mastodon.

So, it's complete for the single section.  I was looking for a place in an existing doc, and was told to just create a new doc for this.

File: Running-Mastodon/Using CDNs.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Using Mastodon Behind A CDN
+Mastodon behind a CDN such as Cloudflare or Akamai, allows you to save bandwidth on your host, by offloading processing elsewhere.  Other benefits with many CDNs are also available, such as dynamic routing, or SSL termination.
+
+While there are benefits to such a setup, it can be very complex, and make issues difficult to troubleshoot.  You will definitley want to have a development/testing instance of Mastodon that is not behind the CDN, in order to test new code deployed to your system.
```
[the remark this answers is not in the document]
Comment (coreyreichle) [reply Running-Mastodon/Using CDNs.md]: fixed

File: Running-Mastodon/Using CDNs.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+## Cloudflare
+Cloudflare is a service that provides SSL termination, DNS hosting, and additional CDN services.  Only the DNS hosting and SSL termination is viable to use with Mastodon.
+
+If you are wanting Cloudflare to terminate your SSL, you will need to open port 80 to the world, and ensure that the CDN option is always on for your domain record.  However, you will need to create page rules to turn off the other acceleration features.
```
[the remark this answers is not in the document]
Comment (coreyreichle) [reply Running-Mastodon/Using CDNs.md]: Added verbiage explaining security implications.

File: Running-Mastodon/Using CDNs.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Using Mastodon Behind A CDN
+Mastodon behind a CDN such as Cloudflare or Akamai, allows you to save bandwidth on your host, by offloading processing elsewhere.  Other benefits with many CDNs are also available, such as dynamic routing, or SSL termination.
```
[the remark this answers is not in the document]
Comment (coreyreichle) [reply Running-Mastodon/Using CDNs.md]: I don't know if it's fluff, but I fleshed it out for those who might be considering using a CDN.

**Comment by nullkal** [comment]
From what I experienced in mstdn.jp, when you want to use CloudFlare for all Mastodon instance you have to care about:

* CloudFlare can proxy websocket connections, but maximum number of concurrent connections is limited; CloudFlare support says the limit is determined by plan, but it is unclear how many connections you can use exactly (Yes, this is the reason why mstdn.jp used to separate streaming endpoint to streaming.mstdn.jp).
* CloudFlare has "Email Address Obfuscation" feature, and it seems that this feature detects "@username@mastodon-instance.example.com" as an email address. So, you have to disable it.
* All features which may rewrite CSS/JS automatically (e.g. Auto Minify, etc...) must be disabled.
* You have to care about caching feature, or your instance will experience serious data breach.

I don't recommend to apply CloudFlare termination for all Mastodon instance. Using CDN to deliver assets and media is enough, almost all of the other data needs to be rendered dynamically and CDN has no merit for them. Regarding to SSL termination, you can use Let's Encrypt, so I highly recommend to use it.

**Comment by nightpool** [comment]
I don't think we should recommend users enable cloudflare, and I don't think this pull request does a good enough job for saying why it should be included.

**Pull request closed without merging**